baseplate/events/publisher.py

```diff
@@ -0,0 +1,203 @@
+        if not self.batch_start:
+            self.batch_start = time.time()
+
+    def flush(self):
+        """Explicitly flush the batch if any items are enqueued."""
+        if self.batch:
+            self.consumer.consume_batch(self.batch)
+            self.batch = []
+            self.batch_size = self.consumer.batch_size_overhead
+            self.batch_start = None
+
+
+def gzip_compress(content):
+    buf = BytesIO()
+    with gzip.GzipFile(fileobj=buf, mode="wb", compresslevel=9) as f:
+        f.write(content)
+    return buf.getvalue()
+
+
+class BatchConsumer(object):
+    batch_size_overhead = 2  # the [] that wrap the json list
+    batch_size_limit = 500 * 1024
+
+    def __init__(self, metrics_client, cfg):
+        self.metrics = metrics_client
+        self.url = "https://%s/v1" % cfg.collector.hostname
+        self.key_name = cfg.key.name
+        self.key_secret = cfg.key.secret
+        self.session = requests.Session()
+
+    @staticmethod
+    def get_item_size(item):
+        # the item is already serialized, so we'll just add one byte for comma
+        return len(item) + 1
+
+    def _sign_payload(self, payload):
+        digest = hmac.new(self.key_secret, payload, hashlib.sha256).hexdigest()
+        return "key={key}, mac={mac}".format(key=self.key_name, mac=digest)
+
+    def consume_batch(self, items):
+        logger.info("sending batch of %d items", len(items))
+        payload = b"[" + b",".join(items) + b"]"
+        compressed_payload = gzip_compress(payload)
+        headers = {
+            "Date": email.utils.formatdate(usegmt=True),
+            "User-Agent": "baseplate-event-publisher/1.0",
+            "Content-Type": "application/json",
+            "X-Signature": self._sign_payload(payload),
+            "Content-Encoding": "gzip",
+        }
+
+        while True:
+            try:
+                with self.metrics.timer("post"):
+                    response = self.session.post(
+                        self.url,
+                        headers=headers,
+                        data=compressed_payload,
+                        timeout=POST_TIMEOUT,
+
+                        # http://docs.python-requests.org/en/latest/user/advanced/#keep-alive
+                        stream=False,
+                    )
+                response.raise_for_status()
+            except requests.HTTPError as exc:
+                self.metrics.counter("error.http").increment()
+                logger.exception("HTTP Request failed")
+
+                # we should crash if it's our fault
+                if getattr(exc, "response", None) and exc.response.status_code < 500:
+                    raise
+            except IOError:
+                self.metrics.counter("error.io").increment()
+                logger.exception("HTTP Request failed")
+            else:
+                self.metrics.counter("sent").increment(len(items))
+                return
+
+            time.sleep(RETRY_DELAY)
```

Comment:
:+1: fwiw, given the current rate of events from r2 servers and the size of the queues we should be good for at least half an hour of consecutive collector downtime before the r2 queues start filling up